src/overlay/RandomPeerSource.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "overlay/RandomPeerSource.h"
+#include "main/Application.h"
+#include "overlay/PeerRecord.h"
+
+#include <algorithm>
+
+namespace stellar
+{
+
+RandomPeerSource::RandomPeerSource(Application& app, size_t batchSize)
+    : mApp(app), mBatchSize(batchSize)
+{
+}
+
+std::vector<PeerRecord>&
+RandomPeerSource::selectPeers(bool outbound)
+{
+    return outbound ? mOutboundPeers : mInboundPeers;
+}
+
+std::vector<PeerRecord>
+RandomPeerSource::getRandomPeers(bool outbound, size_t size,
```

Comment:
I do not like caching of `PeerRecord` like this, it's quite dangerous:
I would rather have a single class managing access to the database that guarantees that we can't override values by accident.

You could still do something like with this class, but the `PeerRecord` used would be shared with the data layer class, and the datalayer class would make sure that we deal with all cases, like if we update a peer record from inbound to outbound for some reason, it would get properly moved from one list to the other.

Actually, thinking about this more: if we have a proper cache for peers, this `getRandomPeers` class probably gets a lot simpler as all it needs to do is ensure that there are enough peers in the cache (if not, force a load), then pick random entries from the cache (partitioned outbound|inbound)

Reply:
I've removed usage of PeerRecord from most of the code, as in reality we only used it to be stored in database. With new PeerManager/PeerRecord the caching does not occur. The only stored values in RandomPeerSource are PeerBareAddress-es.